File: 23.tell-dont-ask/src/typescript/useCase/orderShipmentUseCase.ts
import { OrderRepository } from '../repository/orderRepository';
import { ShipmentService } from '../service/shipmentService';
import { OrderStatus } from '../domain/orderStatus';
import { OrderShipmentRequest } from './orderShipmentRequest';
import { OrderCannotBeShippedException } from './orderCannotBeShippedException';
import { OrderCannotBeShippedTwiceException } from './orderCannotBeShippedTwiceException';

export class OrderShipmentUseCase {
  constructor(
    private readonly orderRepository: OrderRepository,
    private readonly shipmentService: ShipmentService
  ) {}

  run(request: OrderShipmentRequest): void {
    const order = this.orderRepository.getById(request.orderId)!;

    if (order.status === OrderStatus.Created || order.status === OrderStatus.Rejected) {
      throw new OrderCannotBeShippedException();
    }

    if (order.status === OrderStatus.Shipped) {
      throw new OrderCannotBeShippedTwiceException();
    }

    this.shipmentService.ship(order);

    order.status = OrderStatus.Shipped;
    this.orderRepository.save(order);
  }
}

File: 23.tell-dont-ask/src/java/src/main/java/com/telldontask/usecase/OrderShipmentUseCase.java
package com.telldontask.usecase;

import com.telldontask.domain.Order;
import com.telldontask.domain.OrderStatus;
import com.telldontask.repository.OrderRepository;
import com.telldontask.service.ShipmentService;

public class OrderShipmentUseCase {
    private final OrderRepository orderRepository;
    private final ShipmentService shipmentService;

    public OrderShipmentUseCase(OrderRepository orderRepository, ShipmentService shipmentService) {
        this.orderRepository = orderRepository;
        this.shipmentService = shipmentService;
    }

    public void run(OrderShipmentRequest request) {
        Order order = orderRepository.getById(request.getOrderId());

        if (order.getStatus() == OrderStatus.CREATED || order.getStatus() == OrderStatus.REJECTED) {
            throw new OrderCannotBeShippedException();
        }

        if (order.getStatus() == OrderStatus.SHIPPED) {
            throw new OrderCannotBeShippedTwiceException();
        }

        shipmentService.ship(order);

        order.setStatus(OrderStatus.SHIPPED);
        orderRepository.save(order);
    }
}
